### PR TITLE
增加了一些新的函数支持，具体见下方。

### PR DESCRIPTION
**Pull request by iceneve**
在原有工程的基础上修复了一些bug，并新增加了以下函数支持：

coalesce: 可以有任意个数的参数，依照参数的顺序选定结果值。如果第一个参数存在，取其值；否则判断下一个参数；如果都不存在，返回null 。

```sql
SELECT name, coalesce(age2, age1) as myAge from mytest  LIMIT 0, 1
```

min_bw(X, Y) : 取X、Y两个字段中的最小值 max_bw(X, Y) : 取X、Y两个字段中的最大值 。注：bw 是 between 简写 。

```sql
SELECT name, max_bw(age1, age2) as maxAge from mytest  LIMIT 0, 10
```

case_new : 相比已有的case when 可增加多个判断逻辑及提供默认值。 

```sql
select name, case_new(age<=1, 'LOW', age=2, 'SENIOR', age>=3, 'HIGH',  default, '未知1') as myAge from bank  LIMIT 0, 10
```

if：现在支持除 =、!= 以外的二元操作符，如 >、>=、<、<=。

```sql
select  name, if(age <= 18, 'Y', 'O') as myGender from bank  LIMIT 0, 10
```

 percentile_ranks(field, value1, value2,….) ： 百分位统计，统计value值在整个field数值区间中的百分位。

```sql
  SELECT percentile_ranks(age, 3, 8, 12, alias=rankAge) FROM bank
```

movingavg:支持计算指定窗口的移动平均值。注意后面必须加上 GROUP BY date_histogram或dhg（简写。）

```sql
SELECT income / 10 AS myincome, sum(income / 10) AS incomeSum, movingavg(field=incomeSum,window=3,alias=incomeSume_avg) FROM bank3   GROUP BY dhg(field=insert_time, 'interval'=1d, alias=insert_time_dhg) ORDER BY insert_time ASC LIMIT 10
```

rollingavg:支持计算指定窗口的滚动标准差。注意后面必须加上 GROUP BY date_histogram或dhg（简写）。

```sql
SELECT income / 10 AS myincome, sum(income / 10) AS incomeSum, rollingstd(field=incomeSum,window=2,alias=incomeSume_avg) FROM bank3   GROUP BY dhg(field=insert_time, 'interval'=1d, alias=insert_time_dhg) ORDER BY insert_time ASC LIMIT 10
```

### CODE REVIEW

File: src/main/java/org/elasticsearch/plugin/nlpcn/executors/CSVResultsExtractor.java [hunk 1]
```diff
@@ -1,31 +1,28 @@
...
 import org.nlpcn.es4sql.query.DefaultQueryAction;
 import org.nlpcn.es4sql.query.QueryAction;
 
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (hola-holo) [inline src/main/java/org/elasticsearch/plugin/nlpcn/executors/CSVResultsExtractor.java:25]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/nlpcn/es4sql/SQLFunctions.java [hunk 2]
```diff
@@ -3,16 +3,14 @@
...
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import org.elasticsearch.common.collect.Tuple;
-import org.elasticsearch.plugin.nlpcn.executors.CSVResult;
 import org.nlpcn.es4sql.domain.KVValue;
 
-import java.util.List;
-import java.util.Random;
-import java.util.Set;
+import java.util.*;
```
Comment (hola-holo) [inline src/main/java/org/nlpcn/es4sql/SQLFunctions.java:13]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/nlpcn/es4sql/parse/SqlParser.java [hunk 3]
```diff
@@ -12,14 +12,7 @@
...
 import com.alibaba.druid.sql.ast.expr.SQLQueryExpr;
-import com.alibaba.druid.sql.ast.statement.SQLDeleteStatement;
-import com.alibaba.druid.sql.ast.statement.SQLExprTableSource;
-import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource;
-import com.alibaba.druid.sql.ast.statement.SQLSelectGroupByClause;
-import com.alibaba.druid.sql.ast.statement.SQLSelectItem;
-import com.alibaba.druid.sql.ast.statement.SQLSelectOrderByItem;
-import com.alibaba.druid.sql.ast.statement.SQLTableSource;
-import com.alibaba.druid.sql.ast.statement.SQLUnionQuery;
+import com.alibaba.druid.sql.ast.statement.*;
```
Comment (hola-holo) [inline src/main/java/org/nlpcn/es4sql/parse/SqlParser.java:15]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/nlpcn/es4sql/query/AggregationQueryAction.java [hunk 4]
```diff
@@ -11,12 +12,13 @@
 import org.elasticsearch.client.Client;
 import org.elasticsearch.index.query.QueryBuilder;
 import org.elasticsearch.join.aggregations.JoinAggregationBuilders;
-import org.elasticsearch.search.aggregations.AggregationBuilder;
-import org.elasticsearch.search.aggregations.AggregationBuilders;
-import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.search.aggregations.*;
```
Comment (hola-holo) [inline src/main/java/org/nlpcn/es4sql/query/AggregationQueryAction.java:16]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java [hunk 5]
```diff
@@ -28,25 +25,22 @@
...
 
 import java.io.IOException;
 import java.math.BigDecimal;
 import java.time.ZoneOffset;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (hola-holo) [inline src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java:41]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java [hunk 6]
```diff
@@ -28,25 +25,22 @@
...
+import org.elasticsearch.search.aggregations.pipeline.BucketSelectorPipelineAggregationBuilder;
+import org.elasticsearch.search.aggregations.pipeline.MovFnPipelineAggregationBuilder;
 import org.elasticsearch.search.aggregations.support.ValuesSourceAggregationBuilder;
 import org.elasticsearch.search.sort.SortOrder;
 import org.nlpcn.es4sql.Util;
-import org.nlpcn.es4sql.domain.Field;
-import org.nlpcn.es4sql.domain.KVValue;
-import org.nlpcn.es4sql.domain.MethodField;
-import org.nlpcn.es4sql.domain.Where;
+import org.nlpcn.es4sql.domain.*;
```
Comment (hola-holo) [inline src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java:33]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java [hunk 7]
```diff
@@ -8,11 +9,7 @@
 import org.elasticsearch.join.aggregations.JoinAggregationBuilders;
 import org.elasticsearch.script.Script;
 import org.elasticsearch.script.ScriptType;
-import org.elasticsearch.search.aggregations.AbstractAggregationBuilder;
-import org.elasticsearch.search.aggregations.AggregationBuilder;
-import org.elasticsearch.search.aggregations.AggregationBuilders;
-import org.elasticsearch.search.aggregations.BucketOrder;
-import org.elasticsearch.search.aggregations.InternalOrder;
+import org.elasticsearch.search.aggregations.*;
```
Comment (hola-holo) [inline src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java:12]: our static analysis engine detects that your pull request includes a change from an explicit import to a wildcard import. According to the best practice (http://www.javadude.com/articles/importondemandisevil.html), import star (import on demand) is a bad practice, wildcard import can sometimes cause name conflicts and ambiguities.

File: src/main/java/org/elasticsearch/plugin/nlpcn/executors/CSVResultsExtractor.java (hunk 1 above)
Comment (shi-yuan) [reply src/main/java/org/elasticsearch/plugin/nlpcn/executors/CSVResultsExtractor.java:25]: resolved in master

File: src/main/java/org/nlpcn/es4sql/SQLFunctions.java (hunk 2 above)
Comment (shi-yuan) [reply src/main/java/org/nlpcn/es4sql/SQLFunctions.java:13]: resolved in master

File: src/main/java/org/nlpcn/es4sql/parse/SqlParser.java (hunk 3 above)
Comment (shi-yuan) [reply src/main/java/org/nlpcn/es4sql/parse/SqlParser.java:15]: resolved in master

File: src/main/java/org/nlpcn/es4sql/query/AggregationQueryAction.java (hunk 4 above)
Comment (shi-yuan) [reply src/main/java/org/nlpcn/es4sql/query/AggregationQueryAction.java:16]: resolved in master

File: src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java (hunk 5 above)
Comment (shi-yuan) [reply src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java:41]: resolved in master

File: src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java (hunk 6 above)
Comment (shi-yuan) [reply src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java:33]: resolved in master

File: src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java (hunk 7 above)
Comment (shi-yuan) [reply src/main/java/org/nlpcn/es4sql/query/maker/AggMaker.java:12]: resolved in master

**Pull request merged**